File: sdks/go/topics_test.go
package emergent

import (
	"strconv"
	"strings"
	"testing"
)

func TestATopicWithoutAStarIsExact(t *testing.T) {
	for _, topic := range []string{"tick.out", "system.shutdown"} {
		kind, err := ClassifyTopic(topic)
		if err != nil {
			t.Fatalf("ClassifyTopic(%q) returned %v", topic, err)
		}
		if kind != TopicExact {
			t.Errorf("ClassifyTopic(%q) = %q, want %q", topic, kind, TopicExact)
		}
	}
}

func TestATerminalStarIsAPattern(t *testing.T) {
	for _, topic := range []string{"tick.*", "system.started.*", "*", "tick*"} {
		kind, err := ClassifyTopic(topic)
		if err != nil {
			t.Fatalf("ClassifyTopic(%q) returned %v", topic, err)
		}
		if kind != TopicPattern {
			t.Errorf("ClassifyTopic(%q) = %q, want %q", topic, kind, TopicPattern)
		}
	}
}

func TestAStarAnywhereElseIsRejected(t *testing.T) {
	for _, topic := range []string{"*.out", "tick.*.out", "tick.**"} {
		_, err := ClassifyTopic(topic)
		if err == nil {
			t.Fatalf("ClassifyTopic(%q) accepted a topic that can never match", topic)
		}
		if !strings.Contains(err.Error(), topic) {
			t.Errorf("error for %q does not name it: %v", topic, err)
		}
	}
}

func TestTheRejectionMessageNamesAUsableSelector(t *testing.T) {
	_, err := ClassifyTopic("system.*.error")
	if err == nil {
		t.Fatal("expected an error")
	}
	if !strings.Contains(err.Error(), `"system.*"`) {
		t.Errorf("error does not suggest a usable selector: %v", err)
	}
}

func TestAnEmptyTopicIsRejected(t *testing.T) {
	if _, err := ClassifyTopic(""); err == nil {
		t.Fatal("expected an error for an empty topic")
	}
}

func TestAnOversizedPatternIsRejected(t *testing.T) {
	topic := strings.Repeat("a", MaxPatternLen) + "*"
	_, err := ClassifyTopic(topic)
	if err == nil {
		t.Fatal("expected an error for an oversized pattern")
	}
	if !strings.Contains(err.Error(), strconv.Itoa(MaxPatternLen)) {
		t.Errorf("error does not name the limit: %v", err)
	}
}

func TestAPatternExactlyAtTheLimitIsAccepted(t *testing.T) {
	topic := strings.Repeat("a", MaxPatternLen-1) + "*"
	kind, err := ClassifyTopic(topic)
	if err != nil {
		t.Fatalf("ClassifyTopic returned %v", err)
	}
	if kind != TopicPattern {
		t.Errorf("got %q, want %q", kind, TopicPattern)
	}
}

func TestPatternPrefixStripsOnlyATerminalStar(t *testing.T) {
	cases := []struct {
		topic  string
		prefix string
		ok     bool
	}{
		{"tick.*", "tick.", true},
		{"*", "", true},
		{"tick.out", "", false},
		{"*.out", "", false},
		{"a*b*", "", false},
	}
	for _, c := range cases {
		prefix, ok := PatternPrefix(c.topic)
		if prefix != c.prefix || ok != c.ok {
			t.Errorf("PatternPrefix(%q) = (%q, %v), want (%q, %v)", c.topic, prefix, ok, c.prefix, c.ok)
		}
	}
}

func TestTopicMatches(t *testing.T) {
	cases := []struct {
		topic       string
		messageType string
		want        bool
	}{
		{"tick.out", "tick.out", true},
		{"tick.out", "tick.exit", false},
		{"tick.out", "tick.out.deep", false},
		{"tick.*", "tick.out", true},
		{"tick.*", "tick.exit", true},
		{"tick.*", "tick.out.deep", true},
		{"tick.*", "tick", false},
		{"tick.*", "ticker.out", false},
		{"system.started.*", "system.started.ticker", true},
		{"system.started.*", "system.stopped.ticker", false},
		{"*", "tick.out", true},
		{"*", "system.shutdown", true},
		{"*.out", "tick.out", false},
		{"*.out", "*.out", false},
	}
	for _, c := range cases {
		if got := TopicMatches(c.topic, c.messageType); got != c.want {
			t.Errorf("TopicMatches(%q, %q) = %v, want %v", c.topic, c.messageType, got, c.want)
		}
	}
}

func TestPartitioningKeepsOrderAndSeparatesTheTwoKinds(t *testing.T) {
	exact, patterns, err := PartitionTopics([]string{"tick.out", "tick.*", "system.started.*", "a"})
	if err != nil {
		t.Fatalf("PartitionTopics returned %v", err)
	}
	if strings.Join(exact, ",") != "tick.out,a" {
		t.Errorf("exact = %v", exact)
	}
	if strings.Join(patterns, ",") != "tick.*,system.started.*" {
		t.Errorf("patterns = %v", patterns)
	}
}

func TestPartitioningReportsTheFirstUnusableTopic(t *testing.T) {
	_, _, err := PartitionTopics([]string{"tick.out", "a*b", "c*d"})
	if err == nil {
		t.Fatal("expected an error")
	}
	if !strings.Contains(err.Error(), "a*b") {
		t.Errorf("error does not name the first unusable topic: %v", err)
	}
}

func TestPartitioningKeepsAnOverlappingPair(t *testing.T) {
	exact, patterns, err := PartitionTopics([]string{"tick.out", "tick.*"})
	if err != nil {
		t.Fatalf("PartitionTopics returned %v", err)
	}
	if len(exact) != 1 || exact[0] != "tick.out" {
		t.Errorf("exact = %v", exact)
	}
	if len(patterns) != 1 || patterns[0] != "tick.*" {
		t.Errorf("patterns = %v", patterns)
	}
}

func TestIsEmergentMessageType(t *testing.T) {
	cases := map[string]bool{
		"tick.out":                   true,
		"system.started.ticker":      true,
		"with-hyphen_and_underscore": true,
		"SystemEvent":                false,
		"EmergentMessage":            false,
		"":                           false,
		"tick..out":                  false,
		".tick":                      false,
	}
	for messageType, want := range cases {
		if got := IsEmergentMessageType(messageType); got != want {
			t.Errorf("IsEmergentMessageType(%q) = %v, want %v", messageType, got, want)
		}
	}
}

File: sdks/ts/src/topics_test.ts
/**
 * Tests for subscription topic rules.
 * @module
 */

import { assert, assertEquals, assertThrows } from "@std/assert";
import { ValidationError } from "./errors.ts";
import {
  classifyTopic,
  isEmergentMessageType,
  MAX_PATTERN_LEN,
  partitionTopics,
  patternPrefix,
  topicMatches,
} from "./topics.ts";

Deno.test("a topic without a star is exact", () => {
  assertEquals(classifyTopic("tick.out"), "exact");
  assertEquals(classifyTopic("system.shutdown"), "exact");
});

Deno.test("a terminal star is a pattern", () => {
  assertEquals(classifyTopic("tick.*"), "pattern");
  assertEquals(classifyTopic("system.started.*"), "pattern");
  assertEquals(classifyTopic("*"), "pattern");
  assertEquals(classifyTopic("tick*"), "pattern");
});

Deno.test("a star anywhere else is rejected", () => {
  for (const topic of ["*.out", "tick.*.out", "tick.**"]) {
    assertThrows(() => classifyTopic(topic), ValidationError, topic);
  }
});

Deno.test("the rejection message names a usable selector", () => {
  assertThrows(
    () => classifyTopic("system.*.error"),
    ValidationError,
    "'system.*'",
  );
});

Deno.test("an empty topic is rejected", () => {
  assertThrows(() => classifyTopic(""), ValidationError);
});

Deno.test("an oversized pattern is rejected", () => {
  assertThrows(
    () => classifyTopic("a".repeat(MAX_PATTERN_LEN) + "*"),
    ValidationError,
    String(MAX_PATTERN_LEN),
  );
});

Deno.test("a pattern exactly at the limit is accepted", () => {
  assertEquals(classifyTopic("a".repeat(MAX_PATTERN_LEN - 1) + "*"), "pattern");
});

Deno.test("patternPrefix strips only a terminal star", () => {
  assertEquals(patternPrefix("tick.*"), "tick.");
  assertEquals(patternPrefix("*"), "");
  assertEquals(patternPrefix("tick.out"), null);
  assertEquals(patternPrefix("*.out"), null);
  assertEquals(patternPrefix("a*b*"), null);
});

Deno.test("an exact topic matches only itself", () => {
  assert(topicMatches("tick.out", "tick.out"));
  assert(!topicMatches("tick.out", "tick.exit"));
  assert(!topicMatches("tick.out", "tick.out.deep"));
});

Deno.test("a pattern matches every message type with its prefix", () => {
  assert(topicMatches("tick.*", "tick.out"));
  assert(topicMatches("tick.*", "tick.exit"));
  assert(topicMatches("tick.*", "tick.out.deep"));
  assert(!topicMatches("tick.*", "tick"));
  assert(!topicMatches("tick.*", "ticker.out"));
  assert(topicMatches("system.started.*", "system.started.ticker"));
  assert(!topicMatches("system.started.*", "system.stopped.ticker"));
});

Deno.test("a bare star matches everything", () => {
  assert(topicMatches("*", "tick.out"));
  assert(topicMatches("*", "system.shutdown"));
});

Deno.test("a misplaced wildcard matches nothing", () => {
  assert(!topicMatches("*.out", "tick.out"));
  assert(!topicMatches("*.out", "*.out"));
});

Deno.test("partitioning keeps order and separates the two kinds", () => {
  const { exact, patterns } = partitionTopics([
    "tick.out",
    "tick.*",
    "system.started.*",
    "a",
  ]);
  assertEquals(exact, ["tick.out", "a"]);
  assertEquals(patterns, ["tick.*", "system.started.*"]);
});

Deno.test("partitioning reports the first unusable topic", () => {
  assertThrows(
    () => partitionTopics(["tick.out", "a*b", "c*d"]),
    ValidationError,
    "a*b",
  );
});

Deno.test("partitioning keeps an overlapping pair", () => {
  const { exact, patterns } = partitionTopics(["tick.out", "tick.*"]);
  assertEquals(exact, ["tick.out"]);
  assertEquals(patterns, ["tick.*"]);
});

Deno.test("Emergent message types are recognised", () => {
  assert(isEmergentMessageType("tick.out"));
  assert(isEmergentMessageType("system.started.ticker"));
  assert(isEmergentMessageType("with-hyphen_and_underscore"));
});

Deno.test("transport envelope names are not Emergent messages", () => {
  assert(!isEmergentMessageType("SystemEvent"));
  assert(!isEmergentMessageType("EmergentMessage"));
  assert(!isEmergentMessageType(""));
  assert(!isEmergentMessageType("tick..out"));
  assert(!isEmergentMessageType(".tick"));
});
